273601 piano roll 12 tone

blackears opened the pull request:
I've made some changes to the piano roll editor to update the UI to more closely resemble that of other music programs.  I've replaced the transform() scaling which causes scaling artifacts with a system where only the coordinates are translated which allows for sharper drawing of shapes.  I've also given each degree of the scale its own bar line.  Also, it is now possible to click and drag on notes to change their pitch.

Many of the changes are preparation for allowing the user to edit note temporal aspects with the mouse too.

I've uploaded a quick demo:
https://youtu.be/-nsRiEoVfqA

IsaacWeiss commented:
Based on your video, this is a thing of beauty, and I heartily endorse all of your changes.

Jojo-Schmitz commented:
You should squash your commints and have a commit message if "fix #273601: ..." so the corresponding issue gets marked as fixed automatically.
Also you'd need to sign thr [CLA](https://musescire.org/cla)

Jojo-Schmitz commented on .gitignore:
```diff
@@ -43,3 +43,8 @@ MuseScore_General.sf3
+/buildReleaseLog.txt
+/nbproject
+/*.stackdump
+/CMakeLists.txt
```
changes to CMakeLists.txt should surly not get ignored

blackears replied:
I was having some trouble with my submission here.  I needed to modify CMakeLists.txt to get Musescore to compile on my machine.  The change here was an attempt to stop my local CMakeLists.txt from being pushed to the repository.  I do not intend to be changing CMakeLists.txt in central.  I'm new to git and not sure how to do this.

Jojo-Schmitz replied:
Should not be needed, if you need to edit that file to get MuseScore to build, either everyone needs the change you your build envirioment is wrong somehow.

Jojo-Schmitz commented on CMakeLists.txt:
```diff
@@ -717,7 +717,7 @@ add_custom_target(lrelease
+#add_custom_command(OUTPUT ${PROJECT_BINARY_DIR}/plugins
+#   COMMAND genManual ${PROJECT_SOURCE_DIR} ${PROJECT_BINARY_DIR}
+#   DEPENDS genManual
+#   )
```
Why? Where does that manual get generated now?

blackears replied:
Not sure where this came from.  I didn't add it.  Might have been generated by a script or by QtCreator.  I'm not sure how to tell git to ignore my local changes to this file.

Jojo-Schmitz replied:
Got to have been you changing it.
Prevent it by simply by not commint it.

Jojo-Schmitz commented on mscore/events.cpp:
```diff
@@ -847,7 +857,9 @@ void ScoreView::changeState(ViewState s)
       state = s;
+                        //printf("before changeState\n");
       mscore->changeState(mscoreState());
+                        //printf("after changeState\n");
```
better remove all these debugging artifacts

Jojo-Schmitz commented on mscore/piano.cpp:
```diff
@@ -467,5 +470,5 @@ void Piano::setOrientation(PianoOrientation o)
       }
-}
 
+}
```
Only whitespace changes, none of´which is needed

Jojo-Schmitz commented on mscore/pianoroll.cpp:
```diff
@@ -179,9 +176,10 @@ PianorollEditor::PianorollEditor(QWidget* parent)
-      connect(gv,          SIGNAL(magChanged(double,double)),  ruler, SLOT(setMag(double,double)));
-      connect(gv,          SIGNAL(magChanged(double,double)),  piano, SLOT(setMag(double,double)));
+      connect(gv,          SIGNAL(xZoomChanged(qreal)),  ruler, SLOT(setXZoom(qreal)));
+      connect(gv,          SIGNAL(noteHeightChanged(int)),  piano, SLOT(setNoteHeight(int)));
```
allignment got lost

Jojo-Schmitz commented on mscore/pianoroll.cpp:
```diff
@@ -291,6 +309,8 @@ void PianorollEditor::readSettings()
 void PianorollEditor::setXpos(int x)
       {
+      //printf("PianorollEditor::setXpos x:%d\n", x);
+      
```
debugging artifact

Jojo-Schmitz commented on mscore/pianoroll.cpp:
```diff
@@ -498,7 +498,9 @@ void PianorollEditor::heartBeat(Seq* seq)
-                  gv->ensureVisible(tick);
+                  {
+                  gv->scrollToTick(tick);
+                  }
```
unneccessary change

blackears commented:
I'm having trouble figuring to how to squash this.  I'm following the instructions on https://musescore.org/en/handbook/developers-handbook/finding-your-way-around/git-workflow, but every time I run `git rebase master`, I get an error:

D:\dev\github.com\MuseScoreMinGW>git rebase master

It seems that there is already a rebase-apply directory, and
I wonder if you are in the middle of another rebase.  If that is the
case, please try
        git rebase (--continue | --abort | --skip)
If that is not the case, please
        rm -fr "D:/dev/github.com/MuseScoreMinGW/.git/rebase-apply"
and run me again.  I am stopping in case you still have something
valuable there.

```
D:\dev\github.com\MuseScoreMinGW>git rebase --continue
Applying: Updating main piano roll display to have 12 rows per octave.
No changes - did you forget to use 'git add'?
If there is nothing left to stage, chances are that something else
already introduced the same changes; you might want to skip this patch.

Resolve all conflicts manually, mark them as resolved with
"git add/rm <conflicted_files>", then run "git rebase --continue".
You can instead skip this commit: run "git rebase --skip".
To abort and get back to the state before "git rebase", run "git rebase --abort".
```

blackears commented:
I added a few things to my .git/info/exclude and I think the rebase to master worked this time.

anatoly-os commented:
You need to complete squashing first and then `push -f` changes. You can start again with `git rebase -i HEAD~XX`. If git prompts you about other rebasing process, you can cancel it by `git rebase --abort`

blackears commented:
Let me know if everything is squashed correctly.

anatoly-os commented:
You need to do it by yourself :) Check number of commits in this PR. There should be one commit after squashing :) also, you don't need to rebase your changes after squshing on top of master. PS. Don't mix squashing and rebasing in your mind, git commands are similar, but actions are totally different.
Thank you for your patience!

blackears commented:
Okay, I'm becoming really confused here.

I believe I have squashed my changes together using the `git rebase -i ###` command.  However, I cannot push my changes because I get a message saying that the tip of my current branch is behind, even though I have fetched and merged to the master:
```

D:\dev\github.com\MuseScoreMinGW>git checkout master
Switched to branch 'master'
Your branch is ahead of 'origin/master' by 60 commits.
  (use "git push" to publish your local commits)

D:\dev\github.com\MuseScoreMinGW>git merge upstream/master
Already up to date.

D:\dev\github.com\MuseScoreMinGW>git checkout 273601-piano-roll-12-tone
Switched to branch '273601-piano-roll-12-tone'

D:\dev\github.com\MuseScoreMinGW>git rebase master
Current branch 273601-piano-roll-12-tone is up to date.

D:\dev\github.com\MuseScoreMinGW>git push origin 273601-piano-roll-12-tone
To https://github.com/blackears/MuseScore.git
 ! [rejected]            273601-piano-roll-12-tone -> 273601-piano-roll-12-tone (non-fast-forward)
error: failed to push some refs to 'https://github.com/blackears/MuseScore.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
```

anatoly-os commented:
Please, make sure you read our gitflow and understand it (https://musescore.org/en/handbook/developers-handbook/finding-your-way-around/git-workflow). After you make a PR, you cannot merge/rebase revisions. PR is based on some revision and you cannot change it. You need to just squash your commits, so that it will be one based on the same initial revision, and then you can force push changes to the branch in your fork (origin) and this PR will be updated automatically.

PS. I'm a bit wrong about rebasing to master, but it is not important right now. Now, squashing is important.

blackears commented:
It would be helpful if someone added some instructions to that document regarding how to squash commits.  As it is, someone new to git who is following it will be pushing unsquashed commits.

Given that I have merged after the push request, would you know how I would get things back into  shape?  Should I just squash everything into a single commit?

blackears commented:
I've tried doing a hard reset, redoing the rebase and then pushing again, but I'm getting the same error.  The hint tells me to `git pull`, but I don't know which branch to pull or even if I should pull anything given your policy.  Are the commands I used below useful for getting  me back to the right point?  How can I finish this?

```
fbb1e4d3b HEAD@{945}: commit: Now displaying drumset names in PianoKeyboard when relevant.
dec9a3dec HEAD@{946}: commit: Cleaning things up for resubmission.
b0f065cab HEAD@{947}: checkout: moving from bebc557ebf3eadc9d0e584616859caff8a8a1535 to 273601-piano-roll-12-tone
bebc557eb HEAD@{948}: rebase: checkout master
b0f065cab HEAD@{949}: commit: Adding CMakeLists.txt back in.
02d0eedf5 HEAD@{950}: checkout: moving from master to 273601-piano-roll-12-tone
bebc557eb HEAD@{951}: checkout: moving from 273601-piano-roll-12-tone to master
02d0eedf5 HEAD@{952}: commit: Removing CMakeLists.txt
...

D:\dev\github.com\MuseScoreMinGW>git reset --hard fbb1e4d3b
HEAD is now at fbb1e4d3b Now displaying drumset names in PianoKeyboard when relevant.


D:\dev\github.com\MuseScoreMinGW>git rebase -i HEAD~946
unix2dos: converting file D:/dev/github.com/MuseScoreMinGW/.git/rebase-merge/git-rebase-todo to DOS format...
dos2unix: converting file D:/dev/github.com/MuseScoreMinGW/.git/rebase-merge/git-rebase-todo to Unix format...
hint: Waiting for your editor to close the file... unix2dos: converting file D:/dev/github.com/MuseScoreMinGW/.git/COMMIT_EDITMSG to DOS format...
dos2unix: converting file D:/dev/github.com/MuseScoreMinGW/.git/COMMIT_EDITMSG to Unix format...
[detached HEAD 2ab5db8fc] Updating main piano roll display to have 12 rows per octave.
 Date: Sun Jun 24 13:58:41 2018 -0400
 18 files changed, 1753 insertions(+), 420 deletions(-)
 create mode 100644 mscore/pianokeyboard.cpp
 create mode 100644 mscore/pianokeyboard.h
 create mode 100644 mscore/pianoruler.cpp
 create mode 100644 mscore/pianoruler.h
Successfully rebased and updated refs/heads/273601-piano-roll-12-tone.


D:\dev\github.com\MuseScoreMinGW>git push origin 273601-piano-roll-12-tone
To https://github.com/blackears/MuseScore.git
 ! [rejected]            273601-piano-roll-12-tone -> 273601-piano-roll-12-tone (non-fast-forward)
error: failed to push some refs to 'https://github.com/blackears/MuseScore.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.

```

blackears commented:
The force push seems to be checking everything in.  On the plus side, I can still build and run a debug version on my local machine and everything still seems to be there.  Should I just checkout a new branch and start over?

anatoly-os commented:
You did something wrong, indeed... It is better to start new branch, then cherry pick your commits one by one in correct sequence (the oldest ones are the first for cherry picking). Then, perform squashing. After that you will have one commit with all changes. And then you can push this new branch and create new PR.

lasconic commented:
If you need help, don't hesitate to join us on IRC : https://kiwiirc.com/client/irc.freenode.net/?nick=musescore|?#musescore It's easier to fix git issues via chat.

lasconic commented:
And from what I see here I would do the following
```
$ git status
```
Make sure you don't have anything in there, if you do `git add` them and `git commit` them

Then 
````
git checkout master
git fetch upstream
git merge upstream/master
git checkout -b 273601-piano-roll-12-tone-2
git cherry-pick 2ab5db8
git push origin HEAD
```

Then create a PR from the 273601-piano-roll-12-tone-2 branch.

From there, we can guide you to make more changes and fix the gitignore etc...

blackears commented:
After completely fubarring this pull request, I've made a fresh start over on https://github.com/musescore/MuseScore/pull/3784/files

lasconic commented:
Let's close this one !
